security/rbac/engine/engine.go

```diff
@@ -0,0 +1,192 @@
+// +build go1.10
+
+/*
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package engine
+
+import (
+	pb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2"
+	cel "github.com/google/cel-go/cel"
+	expr "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+// AuthorizationArgs is the input of CEL engine.
+type AuthorizationArgs struct {
+	md       metadata.MD
+	peerInfo *peer.Peer
+}
+
+// Converts AuthorizationArgs into the evaluation map for CEL.
+func (args AuthorizationArgs) toEvalMap() map[string]interface{} {
+	// TODO(@ezou): implement the conversion logic.
+	return nil
+}
+
+// Decision is the enum type that represents different authorization
+// decisions a CEL engine can return.
+type Decision int32
+
+const (
+	// DecisionAllow indicates allowing the RPC to go through.
+	DecisionAllow Decision = iota
+	// DecisionDeny indicates denying the RPC from going through.
+	DecisionDeny
+	// DecisionUnknown indicates that there is insufficient information to
+	// determine whether or not an RPC call is authorized.
+	DecisionUnknown
+)
+
+// String returns the string representation of a Decision object.
+func (d Decision) String() string {
+	return [...]string{"DecisionAllow", "DecisionDeny", "DecisionUnknown"}[d]
+}
+
+// AuthorizationDecision is the output of CEL engine.
+// If decision is allow or deny, policyNames will either contain the names of
+// all the policies matched in the engine that permitted the action, or be
+// empty as the decision was made after all conditions evaluated to false.
+// If decision is unknown, policyNames will contain the list of policies that
+// evaluated to unknown.
+type AuthorizationDecision struct {
+	decision    Decision
+	policyNames []string
+}
+
+// Converts an expression to a parsed expression, with SourceInfo nil.
+func exprToParsedExpr(condition *expr.Expr) *expr.ParsedExpr {
+	return &expr.ParsedExpr{Expr: condition}
+}
+
+// Converts an expression to a CEL program.
+func exprToProgram(condition *expr.Expr) *cel.Program {
+	// TODO(@ezou): implement the conversion from expr to CEL program.
+	return nil
```

Comment:
I think you're looking for roughly the following:

```go
// Note: env can be shared across multiple Checks / Program constructions.
env := cel.NewEnv(
   cel.Declarations(... vars ..., ... custom_funcs ...),
)
// pretend cel.ExprToAst exists
ast, iss := env.Check(cel.ExprToAst(condition))
if iss.Err() != nil {
   return nil, iss.Err()
}
// Check that the expression will evaluate to a boolean.
if !proto.Equal(ast.ResultType(), decls.Bool) {
  return nil, fmt.Errorf("expected boolean condition")
}
// Build the program plan.
return env.Program(ast, 
    cel.Functions( ... func_impls ...), 
    cel.EvalOptions(cel.OptOptimize),
)
```

Reply:
I have a rough implementation of this method on a separate branch already, but this seems more detailed than the one I have. Thanks!